### PR TITLE
feat - Support webpack 5 emit

### LINKED ISSUE #40: Use webpack 5 API
It looks like we're getting the following warning against webpack 5:

```
(node:52989) [DEP_WEBPACK_COMPILATION_ASSETS] DeprecationWarning: Compilation.assets will be frozen in future, all modifications are deprecated.
BREAKING CHANGE: No more changes should happen to Compilation.assets after sealing the Compilation.
	Do changes to assets earlier, e. g. in Compilation.hooks.processAssets.
	Make sure to select an appropriate stage from Compilation.PROCESS_ASSETS_STAGE_*.
```

Here's a related comment: https://github.com/webpack/webpack/issues/11425#issuecomment-686606318 .

### PR DESCRIPTION
**Pull request by bebraw**
Closes #40.

Note that I'm testing only against webpack 5 at the moment. I can add webpack 4 tests as well though that will complicate the testing setup a bit (we'll need to differentiate between 4 and 5 in directory structure). If webpack 5 only is enough, then the current solution will do.

When running the tests, you see something like this:

```
(node:13253) [DEP_WEBPACK_JAVASCRIPT_MODULES_PLUGIN] DeprecationWarning: webpack.JavascriptModulesPlugin has moved to webpack.javascript.JavascriptModulesPlugin
(node:13253) [DEP_WEBPACK_LIBRARY_TEMPLATE_PLUGIN] DeprecationWarning: webpack.LibraryTemplatePlugin is deprecated and has been replaced by compilation.outputOptions.library or compilation.addEntry + passing a library option
(node:13253) [DEP_WEBPACK_SINGLE_ENTRY_PLUGIN] DeprecationWarning: SingleEntryPlugin was renamed to EntryPlugin
(node:13253) [DEP_WEBPACK_OPTIONS_DEFAULTER] DeprecationWarning: webpack.WebpackOptionsDefaulter is deprecated and has been replaced by webpack.config.getNormalizedWebpackOptions and webpack.config.applyWebpackOptionsDefaults
```

I'll try to dig where those are coming from

### CODE REVIEW

**Comment by bebraw** [comment]
Here's the full trace:

```
(node:13894) [DEP_WEBPACK_JAVASCRIPT_MODULES_PLUGIN] DeprecationWarning: webpack.JavascriptModulesPlugin has moved to webpack.javascript.JavascriptModulesPlugin
    at get JavascriptModulesPlugin (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/webpack/lib/index.js:181:4)
    at Function.JavascriptModulesPlugin (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/webpack/lib/util/memorize.js:22:13)
    at Object.<anonymous>.__importStar (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/src/index.ts:19:96)
    at Object.<anonymous> (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/src/index.ts:2:1)
    at Runtime._execModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:1245:24)
    at Runtime._loadModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:844:12)
    at Runtime.requireModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:694:10)
    at Runtime.requireModuleOrMock (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:865:21)
    at Object.<anonymous> (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/test/plugin.test.ts:4:1)
    at Runtime._execModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:1245:24)
(node:13894) [DEP_WEBPACK_LIBRARY_TEMPLATE_PLUGIN] DeprecationWarning: webpack.LibraryTemplatePlugin is deprecated and has been replaced by compilation.outputOptions.library or compilation.addEntry + passing a library option
    at get LibraryTemplatePlugin (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/webpack/lib/index.js:191:4)
    at Function.LibraryTemplatePlugin (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/webpack/lib/util/memorize.js:22:13)
    at Object.<anonymous>.__importStar (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/src/index.ts:19:96)
    at Object.<anonymous> (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/src/index.ts:2:1)
    at Runtime._execModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:1245:24)
    at Runtime._loadModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:844:12)
    at Runtime.requireModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:694:10)
    at Runtime.requireModuleOrMock (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:865:21)
    at Object.<anonymous> (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/test/plugin.test.ts:4:1)
    at Runtime._execModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:1245:24)
(node:13894) [DEP_WEBPACK_SINGLE_ENTRY_PLUGIN] DeprecationWarning: SingleEntryPlugin was renamed to EntryPlugin
    at get SingleEntryPlugin (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/webpack/lib/index.js:246:4)
    at Function.SingleEntryPlugin (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/webpack/lib/util/memorize.js:22:13)
    at Object.<anonymous>.__importStar (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/src/index.ts:19:96)
    at Object.<anonymous> (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/src/index.ts:2:1)
    at Runtime._execModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:1245:24)
    at Runtime._loadModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:844:12)
    at Runtime.requireModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:694:10)
    at Runtime.requireModuleOrMock (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:865:21)
    at Object.<anonymous> (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/test/plugin.test.ts:4:1)
    at Runtime._execModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:1245:24)
(node:13894) [DEP_WEBPACK_OPTIONS_DEFAULTER] DeprecationWarning: webpack.WebpackOptionsDefaulter is deprecated and has been replaced by webpack.config.getNormalizedWebpackOptions and webpack.config.applyWebpackOptionsDefaults
    at get WebpackOptionsDefaulter (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/webpack/lib/index.js:271:4)
    at Function.WebpackOptionsDefaulter (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/webpack/lib/util/memorize.js:22:13)
    at Object.<anonymous>.__importStar (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/src/index.ts:19:96)
    at Object.<anonymous> (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/src/index.ts:2:1)
    at Runtime._execModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:1245:24)
    at Runtime._loadModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:844:12)
    at Runtime.requireModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:694:10)
    at Runtime.requireModuleOrMock (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:865:21)
    at Object.<anonymous> (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/test/plugin.test.ts:4:1)
    at Runtime._execModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:1245:24)
```

I wonder if that means we'll have to alter the entry handling as well.

**Comment by bebraw** [comment]
Based on the docs

```
Compilation.entries removed
MIGRATION: Use Compilation.entryDependencies instead
```

Interestingly enough there's no mention of `entryDependencies` in webpack source and the type still has `entries` in it.

I can see `asyncEntrypoints` but that's likely for the entries that return a `Promise`. As we don't support that case yet, we probably don't have to care about this field for now.

File: src/index.ts [hunk 1]
```diff
@@ -158,38 +162,79 @@ function defaultTemplate({
...
 			chunks,
 		} = this.options;
-
 		const files = getFiles(compilation.entrypoints, chunks);
-
 		const options = Object.assign({}, { publicPath }, context, files);
 
 		Promise.resolve((template || defaultTemplate)(options)).then((source) => {
-			compilation.assets[filename] = new RawSource(source);
+			// eslint-disable-next-line
```
Comment (sapegin) [inline src/index.ts:190]: Let's disable a particular rule here, not all of them.

File: src/index.ts [hunk 2]
```diff
@@ -158,38 +162,79 @@ function defaultTemplate({
...
+	private webpack5plugin(compilation: Compilation) {
+		const {
+			filename = 'index.html',
+			publicPath = '',
+			template,
+			context,
+			chunks,
+		} = this.options;
+		const files = getFiles(compilation.entrypoints, chunks);
+		const options = Object.assign({}, { publicPath }, context, files);
```
Comment (sapegin) [inline src/index.ts:207]:
```suggestion
		const options = { publicPath, ...context, ...files};
```

File: test/__snapshots__/plugin.test.ts.snap [hunk 3]
```diff
@@ -9,7 +9,7 @@ exports[`additional body 1`] = `
       
     </head>
     <body>
-      <div>Demo</div><script src=\\"runtime~main.js\\"></script><script src=\\"main.js\\"></script>
+      <div>Demo</div><script src=\\"main.js\\"></script>
```
Comment (sapegin) [inline test/__snapshots__/plugin.test.ts.snap:12]: Are these changes intentional?

**Review by sapegin — COMMENTED** [review]
Overall looks good to me! I've left a few minor comments.

File: test/__snapshots__/plugin.test.ts.snap (hunk 3 above)
Comment (bebraw) [reply test/__snapshots__/plugin.test.ts.snap:12]: Yes, they have to do with webpack 5 default behavior as there the runtime is included to the output bundles by default.

**Comment by bebraw** [comment]
I addressed the comments. I think I'll add a standalone example so we can see if the deprecation warnings appear there as well.

**Review by sapegin — APPROVED** [review]
🧇

**Comment by bebraw** [comment]
I think we still have to solve those deprecation warnings before we can merge but it's starting to get there.

**Comment by bebraw** [comment]
I figured it out. The issue was that for whatever reason, **mini-css-webpack-plugin** was pulling webpack's old type definitions to the project. Now the type definitions are included to webpack itself.

To solve, I simplified the tests a bit and dropped **mini-css-webpack-plugin**. Let's run with this.
